framework/source/class/qx/ui/form/VirtualSelectBox.js

```diff
@@ -445,6 +467,150 @@ qx.Class.define("qx.ui.form.VirtualSelectBox",
       var d = event.getData();
       var old = (d.removed.length ? d.removed[0] : null);
       this.fireDataEvent("changeValue", d.added[0], old);
+    },
+
+
+    /*
+    ---------------------------------------------------------------------------
+      INCREMENTAL SEARCH
+    ---------------------------------------------------------------------------
+    */
+
+    __filterValue : null,
+    __lastMatch : '',
+    __filterUpdateRunning : 0,
+    __filterInput : null,
+    __highlightStyle : null,
+    __lastRich : false,
+
+
+     __addFilterInput : function()
+    {
+      var input = this.__filterInput = new qx.ui.form.TextField().set({
+          appearance : 'widget',
+          liveUpdate : true,
+          height     : 0,
+          width      : 1 // must be > 0
+      });
+      this._add(input);
+      input.addListener("focus", function(e) {
+          // reopen after focus loss
+          this.open();
+      }, this);
+      this.addListener("focus", function(e) {
+          // move focus into TextField
+          input.focus();
+      }, this);
+      input.addListener('changeValue', function(e) {
+          if (this.__filterUpdateRunning === 0) this.__updateDelegate();
+      }, this);
+      input.addListener("keypress", function(e) {
+          switch (e.getKeyIdentifier()) {
+          case 'Enter':
+          case 'Escape':
+              this.__updateDelegate();
+              break;
+          // prevent cursor from being moved to beginning/end of input field
+          case 'Down':
+          case 'Up':
+              e.preventDefault();
+              break;
+          }
+      }, this);
+    },
+
+    __applyMarkHighlightStyle : function(value, old) {
+      var styles = [];
+      if (value != null) {
+        var keys = Object.keys(value);
+        var k;
+        for (k=0; k< keys.length; k++) {
+          var key = keys[k].replace(/[A-Z]/g, m => "-" + m.toLowerCase());
+          styles.push(key + ':' + value[keys[k]]);
+        }
+        this.__highlightStyle = styles.join(';');
+      }
+      else {
+        this.__highlightStyle = '';
+      }
+    },
+
+    _highlightFilterValue : function (item)
+    {
+      return this.__highlightStyle ? item.replace(item, '<span style="' + this.__highlightStyle + '">' + this.__filterValue+'</span>') : item;
+    },
+
+    __updateDelegate : function(lastFilterValue)
+    {
+      this.__filterUpdateRunning++;
+      var filterValue = (lastFilterValue !== undefined) ? lastFilterValue : this.__filterInput.getValue();
+      this.__filterValue = filterValue;
+      // create and apply new filter
+      var delegate = {
+          filter : function(item) {
+              return item.match(filterValue);
+          },
+          configureItem : function(item) {
+              item.setRich(true);
```

Comment:
Does this potentially allow for unanticipated injection of evil code? I understand that `rich` is a prerequisite to attaining this behavior, but wondering if we need a warning someplace that it will be automatically made `rich`? Since the search is always from the beginning of the list item value, maybe highlighting should be an option since the same early text will be highlighted in every row so is pretty but not absolutely necessary?

Reply:
true, the input would have to be escaped html escaped ...

Reply:
> Does this potentially allow for unanticipated injection of evil code? I understand that `rich` is a prerequisite to attaining this behavior, but wondering if we need a warning someplace that it will be automatically made `rich`? 

> Since the search is always from the beginning of the list item value, maybe highlighting should be an option since the same early text will be highlighted in every row so is pretty but not absolutely necessary? And if we don't highlight, then `rich` isn't necessary.
The search is intentionally **not** required to be from the beginning of the string. My use case is a select box containing addresses as items and always searching from the beginning is not very useful, especially of the item starts with a zip code or a town, because at least in our customer's application the street is much more likely to be "relatively" unique.
And thus the highlighting is not just pretty but necessary for the user to recognize why an item matches.
<img width="143" alt="Screenshot 2021-01-08 at 20 27 25" src="https://user-images.githubusercontent.com/198038/104056041-1918f900-51f0-11eb-858b-cf38a1c66ad1.png">

Reply:
> true, the input would have to be escaped html escaped ...

Well, the only HTML that could be entered is what's already there in the item labels. Is this really an issue?

Reply:
But I added `escapeHtml()` around the filter value.

Reply:
the problem is that the list labels are normally not 'rich' ... escapeHtml should take care of the problem

Reply:
> the problem is that the list labels are normally not 'rich' ... escapeHtml should take care of the problem

That would break the use of HTML formatted item labels together with incremental search.

Reply:
Fixed, I think.